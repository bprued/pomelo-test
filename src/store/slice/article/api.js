import { createAsyncThunk } from '@reduxjs/toolkit'
import request, { GETOption } from '../../../utils/request'
import { ARTICLE_URL, APP_KEY } from '../../../utils/api-const' 

export const getMostPopularArticle = createAsyncThunk(
  'article/getMostPopularArticle',
  async (period) => {
    const response = await request(`${ARTICLE_URL}/${period}.json?${APP_KEY}`, GETOption())
    return response
  }
)