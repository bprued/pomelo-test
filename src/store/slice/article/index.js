import { createSlice } from '@reduxjs/toolkit'
import { getMostPopularArticle } from './api'

export const ArticleSlice = createSlice({
  name: 'article',
  initialState: {
    articleList: {
      status: 'idle',
      data: []
    }
  },
  extraReducers: {
    [getMostPopularArticle.pending]: (state) => {
      state.articleList.status = 'loading'
    },
    [getMostPopularArticle.fulfilled]: (state, action) => {
      state.articleList.status = 'succeeded'
      state.articleList.data = action.payload
    },
    [getMostPopularArticle.rejected]: (state) => {
      state.articleList.status = 'failed'
      state.articleList.data = []
    },
  }
})

export const selectArticleList = (state) => state.article.articleList

export default ArticleSlice.reducer
