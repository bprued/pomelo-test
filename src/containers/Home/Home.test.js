import { render, cleanup, fireEvent } from '@testing-library/react'
import HomePage from '.'
import { selectArticleList } from '../../store/slice/article'

jest.mock('react-redux', () => ({
  useSelector: (cb) => cb(),
  useDispatch: () => jest.fn()
}))

jest.mock('../../store/slice/article', () => ({
  selectArticleList: jest.fn()
}))

const setSelectedArticleFn = jest.fn()

const HomePageComponent = (
  <HomePage setSelectedArticle={setSelectedArticleFn} />
)

let articleListState = { 
  status: 'status',
  data: [] 
}
selectArticleList.mockReturnValue(articleListState)

describe('Home Page', () => {
  beforeEach(() => {
    selectArticleList.mockReturnValue(articleListState)
  })
  afterEach(() => {
    cleanup
    selectArticleList.mockReturnValue(articleListState)
    jest.clearAllMocks()
  })
  test('Should match snapshot', () => {
    const { asFragment } = render(HomePageComponent)
    expect(asFragment).toMatchSnapshot()
  })
  test('Should show loader when get list not succeeded', () => {
    const { getByTestId } = render(HomePageComponent)
    expect(getByTestId('loader')).toBeInTheDocument()
  })
  test('Should show article list when get list succeeded', () => {
    articleListState = { 
      status: 'succeeded',
      data: {
        results: [
          {
            id: '01',
            title: 'test',
            abstract: 'test article'
          }
        ]
      } 
    }
    selectArticleList.mockReturnValue(articleListState)
    const { getByText } = render(HomePageComponent)
    expect(getByText('test')).toBeInTheDocument()
  })
  test('Should call setSelectedArticle when click on article', () => {
    let mockArticle = {
      id: '01',
      title: 'test',
      abstract: 'test article'
    }
    articleListState = { 
      status: 'succeeded',
      data: {
        results: [mockArticle]
      } 
    }
    selectArticleList.mockReturnValue(articleListState)
    const { getByText } = render(HomePageComponent)
    fireEvent.click(getByText('test'))
    expect(setSelectedArticleFn).toHaveBeenLastCalledWith(mockArticle)
  })
  test('Should show not found text when search not found', () => {
    const { getByText, getByRole } = render(HomePageComponent)
    fireEvent.change(getByRole('textbox'),{ target: { value: 'test message' } })
    expect(getByText('No Articles Found')).toBeInTheDocument()
  })
})