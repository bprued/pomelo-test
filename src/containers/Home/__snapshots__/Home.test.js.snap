// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Page Should match snapshot 1`] = `[Function]`;
